### PR TITLE
refactor: data scratch file references

### LINKED ISSUE #1759: Data scratch - file reference
# Plan

According to the specification described in epic [#1727](https://github.com/Flank/flank/issues/1727)
add abstraction and implementation for the following part of the data layer:

## Abstraction

`ftl/data/FileReference.kt`

```kotlin
package ftl.data

val downloadFileReference: FileReference.Download get() = TODO()
val uploadFileReference: FileReference.Download get() = TODO()
val existFileReference: FileReference.Exist get() = TODO()

data class FileReference(
    val local: String = "",
    val remote: String = ""
) {
    interface Download : (FileReference) -> FileReference
    interface Upload : (FileReference) -> FileReference
    interface Exist : (FileReference) -> Boolean
}
```

## Target

* `FileReference.kt`

## Adapter

`ftl/adapter/GoogleFileReference.kt`

### Note

Consider reuse or expand `RemoteStorage` implementation.

### PR DESCRIPTION
**Pull request by Sloox**
Fixes #1759 

- Code is refactored according to the description in #1759 
- File reference has been replaced and is acting as expected.

## Checklist

- [x] Unit tested

### CODE REVIEW

**Comment by Sloox** [comment]
@flank-it

**Review by pawelpasterz — APPROVED** [review]

File: test_runner/src/main/kotlin/ftl/adapter/GcStorageDownload.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    (FileReference, Boolean, Boolean) -> FileReference by { fileReference, ifNeeded, ignoreErrors ->
+        if (!ignoreErrors) {
+            if (fileReference.local.isBlank() && fileReference.remote.isBlank()) throw FlankGeneralError("Cannot create empty FileReference")
+        }
+        if (ifNeeded) {
+            if (fileReference.local.isNotBlank()) fileReference
+            else gcStorageDownload(fileReference.remote, ignoreErrors).toApiModel(fileReference)
+        } else {
+            gcStorageDownload(fileReference.remote, ignoreErrors).toApiModel(fileReference)
+        }
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/adapter/GcStorageDownload.kt:19]: could you make `when` instead of nested `if`s?

Comment (Sloox) [reply test_runner/src/main/kotlin/ftl/adapter/GcStorageDownload.kt:19]: 👍

Comment (Sloox) [reply test_runner/src/main/kotlin/ftl/adapter/GcStorageDownload.kt:20]: modified :)

**Review by jan-gogo — CHANGES_REQUESTED** [review]
The domain code cannot have direct references to the API adapters.

**Comment by Sloox** [comment]
@flank-it

**Pull request merged**